cmd/astro-wasm/astro-wasm.go

```diff
@@ -77,47 +87,97 @@ type TransformResult struct {
-	source := jsString(args[0])
-	hash := astro.HashFromSource(source)
-	transformOptions := makeTransformOptions(js.Value(args[1]), hash)
-
-	var doc *astro.Node
-
-	if transformOptions.As == "document" {
-		docNode, _ := astro.Parse(strings.NewReader(source))
-		doc = docNode
-	} else if transformOptions.As == "fragment" {
-		nodes, _ := astro.ParseFragment(strings.NewReader(source), &astro.Node{
-			Type:     astro.ElementNode,
-			Data:     atom.Body.String(),
-			DataAtom: atom.Body,
-		})
-		doc = &astro.Node{
-			Type: astro.DocumentNode,
-		}
-		for i := 0; i < len(nodes); i++ {
-			n := nodes[i]
-			doc.AppendChild(n)
-		}
+// This is spawned as a goroutine to preprocess style nodes using an async function passed from JS
+func preprocessStyle(i int, style *astro.Node, transformOptions transform.TransformOptions, cb func()) {
+	defer cb()
+	attrs := transform.GetAttrs(style)
+	data, _ := wasm_utils.Await(transformOptions.PreprocessStyle.Invoke(style.FirstChild.Data, attrs))
+	str := jsString(data[0])
+	if str == "" {
+		return
 	}
+	style.FirstChild.Data = str
+}
 
-	transform.Transform(doc, transformOptions)
-
-	result := printer.PrintToJS(source, doc, transformOptions)
-
-	switch transformOptions.SourceMap {
-	case "external":
-		return createExternalSourceMap(source, result, transformOptions)
-	case "both":
-		return createBothSourceMap(source, result, transformOptions)
-	case "inline":
-		return createInlineSourceMap(source, result, transformOptions)
-	}
+func Transform() interface{} {
+	return js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		source := jsString(args[0])
+		hash := astro.HashFromSource(source)
+		transformOptions := makeTransformOptions(js.Value(args[1]), hash)
+
+		handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+			resolve := args[0]
+
+			go func() {
+				var doc *astro.Node
+
+				if transformOptions.As == "document" {
+					docNode, _ := astro.Parse(strings.NewReader(source))
+					doc = docNode
+				} else if transformOptions.As == "fragment" {
+					nodes, _ := astro.ParseFragment(strings.NewReader(source), &astro.Node{
+						Type:     astro.ElementNode,
+						Data:     atom.Body.String(),
+						DataAtom: atom.Body,
+					})
+					doc = &astro.Node{
+						Type: astro.DocumentNode,
+					}
+					for i := 0; i < len(nodes); i++ {
+						n := nodes[i]
+						doc.AppendChild(n)
+					}
+				}
+
+				// Hoist styles and scripts to the top-level
+				transform.ExtractScriptsAndStyles(doc)
+
+				// Pre-process styles
+				// Important! These goroutines need to be spawned from this file or they don't work
+				var wg sync.WaitGroup
+				if len(doc.Styles) > 0 {
+					if transformOptions.PreprocessStyle.IsUndefined() != true {
+						for i, style := range doc.Styles {
+							wg.Add(1)
+							i := i
+							go preprocessStyle(i, style, transformOptions, wg.Done)
+						}
+					}
+				}
+				// Wait for all the style goroutines to finish
+				wg.Wait()
```

Comment:
This spawns workers that run the async `preprocessStyle` functions and waits until they are all complete to continue.